README.md
# ms_dart

[![Build Status](https://travis-ci.org/chrisvasquez/ms_dart.svg?branch=master)](https://travis-ci.org/chrisvasquez/ms_dart)
[![Join the community on Spectrum](https://withspectrum.github.io/badge/badge.svg)](https://spectrum.chat/users/chrisvasquez)

Use this package to easily convert various time formats to milliseconds.

## Install

Add in `pubspec.yaml`

```yaml
dependencies:
  ms_dart: 1.0.4
```

## Examples

```dart
MS.toMs('2 days')  // 172800000
MS.toMs('1d')      // 86400000
MS.toMs('10h')     // 36000000
MS.toMs('2.5 hrs') // 9000000
MS.toMs('2h')      // 7200000
MS.toMs('1m')      // 60000
MS.toMs('5s')      // 5000
MS.toMs('1y')      // 31557600000
MS.toMs('100')     // 100
MS.toMs('-3 days') // -259200000
MS.toMs('-1h')     // -3600000
MS.toMs('-200')    // -200
```

### Convert from Milliseconds

```dart
MS.fromMs(60000)             // "1m"
MS.fromMs(2 * 60000)         // "2m"
MS.fromMs(-3 * 60000)        // "-3m"
MS.fromMs(ms('10 hours'))    // "10h"
```

### Time Format Written-Out

```dart
MS.fromMs(60000, { long: true })                // "1 minute"
MS.fromMs(2 * 60000, { long: true })            // "2 minutes"
MS.fromMs(-3 * 60000, { long: true })           // "-3 minutes"
MS.fromMs(MS.toMs('10 hours'), { long: true })  // "10 hours"
```

## Features

- Works both in dart vm, flutter, and in the browser
- If a number is supplied to `MS.fromMS()`, a string with a unit is returned
- If a string that contains the number is supplied to `MS.toMs()`, it returns it as a number (e.g.: it returns `100` for `'100'`)
- If you pass a string with a number and a valid unit to `MS.toMs()`, the number of equivalent milliseconds is returned

## Related Packages

- [ms](https://https://github.com/zeit/ms) - Based on the `ms` javascript package.

## Caught a Bug?

1. [Fork](https://help.github.com/articles/fork-a-repo/) this repository to your own GitHub account and then [clone](https://help.github.com/articles/cloning-a-repository/) it to your local device
2. You can run the tests using: `dart test/ms_dart_test_dart`


CHANGELOG.md
## 1.0.0

- Initial version, created by Stagehand

## 1.0.1

- minor edit

## 1.0.2

- description in pubspec.yaml modified

## 1.0.3

- fix length of description in pubspec.yaml

## 1.0.3-a

- fix of README.md - documentation

## 1.0.4

- fix of version


test/ms_dart_test.dart
import 'package:ms_dart/ms_dart.dart';
import 'package:test/test.dart';

void main() {
  // strings

  group('MS.toMs(string)', () {
    test('should not throw an error', () {
      expect(() {
        MS.toMs('1m');
      }, returnsNormally);
    });

    test('should preserve ms', () {
      expect(MS.toMs('100'), 100);
    });

    test('should convert from m to ms', () {
      expect(MS.toMs('1m'), 60000);
    });

    test('should convert from h to ms', () {
      expect(MS.toMs('1h'), 3600000);
    });

    test('should convert d to ms', () {
      expect(MS.toMs('2d'), 172800000);
    });

    test('should convert w to ms', () {
      expect(MS.toMs('3w'), 1814400000);
    });

    test('should convert s to ms', () {
      expect(MS.toMs('1s'), 1000);
    });

    test('should convert ms to ms', () {
      expect(MS.toMs('100ms'), 100);
    });

    test('should work with decimals', () {
      expect(MS.toMs('1.5h'), 5400000);
    });

    test('should work with multiple spaces', () {
      expect(MS.toMs('1   s'), 1000);
    });

    test('should return NaN if invalid', () {
      expect(MS.toMs('☃').isNaN, true);
      expect(MS.toMs('10-.5').isNaN, true);
    });

    test('should be case-insensitive', () {
      expect(MS.toMs('1.5H'), 5400000);
    });

    test('should work with numbers starting with .', () {
      expect(MS.toMs('.5ms'), 0.5);
    });

    test('should work with negative integers', () {
      expect(MS.toMs('-100ms'), -100);
    });

    test('should work with negative decimals', () {
      expect(MS.toMs('-1.5h'), -5400000);
      expect(MS.toMs('-10.5h'), -37800000);
    });

    test('should work with negative decimals starting with "."', () {
      expect(MS.toMs('-.5h'), -1800000);
    });
  });

  // long strings

  group('MS.toMs(long string)', () {
    test('should not throw an error', () {
      expect(() {
        MS.toMs('53 milliseconds');
      }, returnsNormally);
    });

    test('should convert milliseconds to ms', () {
      expect(MS.toMs('53 milliseconds'), 53);
    });

    test('should convert msecs to ms', () {
      expect(MS.toMs('17 msecs'), 17);
    });

    test('should convert sec to ms', () {
      expect(MS.toMs('1 sec'), 1000);
    });

    test('should convert from min to ms', () {
      expect(MS.toMs('1 min'), 60000);
    });

    test('should convert from hr to ms', () {
      expect(MS.toMs('1 hr'), 3600000);
    });

    test('should convert days to ms', () {
      expect(MS.toMs('2 days'), 172800000);
    });

    test('should work with decimals', () {
      expect(MS.toMs('1.5 hours'), 5400000);
    });

    test('should work with negative integers', () {
      expect(MS.toMs('-100 milliseconds'), -100);
    });

    test('should work with negative decimals', () {
      expect(MS.toMs('-1.5 hours'), -5400000);
    });

    test('should work with negative decimals starting with "."', () {
      expect(MS.toMs('-.5 hr'), -1800000);
    });
  });
  group('MS.fromMs(number, long : true)', () {
    test('should not throw an error', () {
      expect(() {
        MS.fromMs(500.0, long: true);
      }, returnsNormally);
    });

    test('should support milliseconds', () {
      expect(MS.fromMs(500.0, long: true), '500 ms');

      expect(MS.fromMs(-500.0, long: true), '-500 ms');
    });

    test('should support seconds', () {
      expect(MS.fromMs(1000.0, long: true), '1 second');
      expect(MS.fromMs(1200.0, long: true), '1 second');
      expect(MS.fromMs(10000.0, long: true), '10 seconds');

      expect(MS.fromMs(-1000.0, long: true), '-1 second');
      expect(MS.fromMs(-1200.0, long: true), '-1 second');
      expect(MS.fromMs(-10000.0, long: true), '-10 seconds');
    });

    test('should support minutes', () {
      expect(MS.fromMs(60 * 1000.0, long: true), '1 minute');
      expect(MS.fromMs(60 * 1200.0, long: true), '1 minute');
      expect(MS.fromMs(60 * 10000.0, long: true), '10 minutes');

      expect(MS.fromMs(-1 * 60 * 1000.0, long: true), '-1 minute');
      expect(MS.fromMs(-1 * 60 * 1200.0, long: true), '-1 minute');
      expect(MS.fromMs(-1 * 60 * 10000.0, long: true), '-10 minutes');
    });

    test('should support hours', () {
      expect(MS.fromMs(60 * 60 * 1000.0, long: true), '1 hour');
      expect(MS.fromMs(60 * 60 * 1200.0, long: true), '1 hour');
      expect(MS.fromMs(60 * 60 * 10000.0, long: true), '10 hours');

      expect(MS.fromMs(-1 * 60 * 60 * 1000.0, long: true), '-1 hour');
      expect(MS.fromMs(-1 * 60 * 60 * 1200.0, long: true), '-1 hour');
      expect(MS.fromMs(-1 * 60 * 60 * 10000.0, long: true), '-10 hours');
    });

    test('should support days', () {
      expect(MS.fromMs(24 * 60 * 60 * 1000.0, long: true), '1 day');
      expect(MS.fromMs(24 * 60 * 60 * 1200.0, long: true), '1 day');
      expect(MS.fromMs(24 * 60 * 60 * 10000.0, long: true), '10 days');

      expect(MS.fromMs(-1 * 24 * 60 * 60 * 1000.0, long: true), '-1 day');
      expect(MS.fromMs(-1 * 24 * 60 * 60 * 1200.0, long: true), '-1 day');
      expect(MS.fromMs(-1 * 24 * 60 * 60 * 10000.0, long: true), '-10 days');
    });

    test('should round', () {
      expect(MS.fromMs(234234234, long: true), '3 days');

      expect(MS.fromMs(-234234234, long: true), '-3 days');
    });
  });

  group('MS.fromMs(number)', () {
    test('should not throw an exception', () {
      expect(() => MS.fromMs(500), returnsNormally);
    });
    test('should support milliseconds', () {
      expect(MS.fromMs(500), '500ms');

      expect(MS.fromMs(-500), '-500ms');
    });
    test('should support seconds', () {
      expect(MS.fromMs(1000), '1s');
      expect(MS.fromMs(10000), '10s');

      expect(MS.fromMs(-1000), '-1s');
      expect(MS.fromMs(-10000), '-10s');
    });

    test('should support minutes', () {
      expect(MS.fromMs(60 * 1000.0), '1m');
      expect(MS.fromMs(60 * 10000.0), '10m');

      expect(MS.fromMs(-1 * 60 * 1000.0), '-1m');
      expect(MS.fromMs(-1 * 60 * 10000.0), '-10m');
    });

    test('should support hours', () {
      expect(MS.fromMs(60 * 60 * 1000.0), '1h');
      expect(MS.fromMs(60 * 60 * 10000.0), '10h');

      expect(MS.fromMs(-1 * 60 * 60 * 1000.0), '-1h');
      expect(MS.fromMs(-1 * 60 * 60 * 10000.0), '-10h');
    });

    test('should support days', () {
      expect(MS.fromMs(24 * 60 * 60 * 1000.0), '1d');
      expect(MS.fromMs(24 * 60 * 60 * 10000.0), '10d');

      expect(MS.fromMs(-1 * 24 * 60 * 60 * 1000.0), '-1d');
      expect(MS.fromMs(-1 * 24 * 60 * 60 * 10000.0), '-10d');
    });

    test('should round', () {
      expect(MS.fromMs(234234234), '3d');

      expect(MS.fromMs(-234234234), '-3d');
    });
  });
}


lib/src/ms_dart_base.dart
/// Helpers.
const double _s = 1000;
const double _m = _s * 60;
const double _h = _m * 60;
const double _d = _h * 24;
const double _w = _d * 7;
const double _y = _d * 365.25;

/// Class that exposes static methoids [fromMs()] and [toMs()].
class MS {
  /// Convert from milliseconds to string representation of time
  ///
  /// Converts [ms], the number of milliseconds to a string representation
  /// rounded to the closes smallest measure of time.
  /// example:
  /// ```dart
  /// MS.fromMs(6499); // '6s'
  /// MS.fromMs(6500); // '7s'
  /// MS.fromMs(6501); // '7s'
  /// MS.fromMs(6999); // '7s'
  /// ```
  /// [long] determines the representation of time measure, if [long] is `true`
  /// the value above will be '6 seconds'.
  static String fromMs(double ms, {bool long = false}) {
    if (_isFinite(ms)) {
      return long ? _fmtLong(ms) : _fmtShort(ms);
    } else {
      throw Exception('val is not a valid number. val=${ms}');
    }
  }

  /// Convert a string representation fo time to the number of milliseconds
  ///
  /// example:
  /// ```dart
  /// MS.toMs('2 days'); // 172800000
  /// ```
  static double toMs(String val) {
    return _parse(val);
  }
}

/// Check if [s] if not `infinite` or `double.nan`.
bool _isFinite(double s) {
  return (s == s) && ((s == 0) || (s != 2 * s));
}

/// Parse the given `str` and return its value in milliseconds.
double _parse(String str) {
  if (str.length > 100) {
    throw Exception(
      'Too long an argument (> 100 string length ) : len=${str.length}',
    );
  }
  final exp = RegExp(
    r'^(-?(?:\d+)?\.?\d+) *(milliseconds?|msecs?|ms|seconds?|secs?|s|minutes?|mins?|m|hours?|hrs?|h|days?|d|weeks?|w|years?|yrs?|y)?$',
    caseSensitive: false,
    unicode: true,
    dotAll: true,
    multiLine: true,
  );
  final matches = exp.allMatches(str);
  if (matches.isEmpty) {
    return double.nan;
  }
  final n = double.parse(matches.first.group(1));
  final type = (matches.first.group(2) ?? 'ms').toLowerCase();
  switch (type) {
    case 'years':
    case 'year':
    case 'yrs':
    case 'yr':
    case 'y':
      return (n * _y);
    case 'weeks':
    case 'week':
    case 'w':
      return (n * _w);
    case 'days':
    case 'day':
    case 'd':
      return (n * _d);
    case 'hours':
    case 'hour':
    case 'hrs':
    case 'hr':
    case 'h':
      return (n * _h);
    case 'minutes':
    case 'minute':
    case 'mins':
    case 'min':
    case 'm':
      return (n * _m);
    case 'seconds':
    case 'second':
    case 'secs':
    case 'sec':
    case 's':
      return (n * _s);
    case 'milliseconds':
    case 'millisecond':
    case 'msecs':
    case 'msec':
    case 'ms':
      return n;
    default:
      return 0.0;
  }
}

/// Short format for `MS.fromMs()`.
String _fmtShort(double ms) {
  final msAbs = ms.abs();
  if (msAbs >= _d) {
    final result = ms / _d;
    return '${result.round().toInt()}d';
  }
  if (msAbs >= _h) {
    final result = ms / _h;
    return '${result.round().toInt()}h';
  }
  if (msAbs >= _m) {
    final result = ms / _m;
    return '${result.round().toInt()}m';
  }
  if (msAbs >= _s) {
    final result = ms / _s;
    return '${result.round().toInt()}s';
  }
  return '${ms.toInt()}ms';
}

/// Long format for `MS.fromMs()`.
String _fmtLong(double ms) {
  final msAbs = ms.abs();
  if (msAbs >= _d) {
    return _plural(ms, msAbs, _d, 'day');
  }
  if (msAbs >= _h) {
    return _plural(ms, msAbs, _h, 'hour');
  }
  if (msAbs >= _m) {
    return _plural(ms, msAbs, _m, 'minute');
  }
  if (msAbs >= _s) {
    return _plural(ms, msAbs, _s, 'second');
  }
  return '${ms.toInt()} ms';
}

/// Pluralization helper.
String _plural(double ms, double msAbs, double n, String name) {
  final isPlural = msAbs >= n * 1.5;
  final time = ms / n;
  final suffix = isPlural ? 's' : '';
  return '${time.round().toInt()} $name$suffix';
}
